### PR TITLE
Issue #522: Solution: Attempt pcap_create() first

### PR DESCRIPTION
**Pull request by bijalthanawala**
This addresses issue:
**Numeric adapter names can't be used for capture**
(https://github.com/the-tcpdump-group/tcpdump/issues/522)

- Attempt pcap_create() first and only on failure numeric indexes are scanned
- Handle use-case: Do not do the above when interface  (-i) was not specified by the user
- Handle special case  (negative zero: -i -0)

Todo / Question:
Do we need to modify tcpdump man page to explain this behaviour ?

### CODE REVIEW

**Comment by nbastin** [comment]
This is sortof pedantic, but I think important.  Can we remove the comment that says:

```
+	 * This should be OK on UNIX systems, as interfaces
+	 * shouldn't have names that begin with digits.
```

This is an opinion, and has no basis in the POSIX specification or AT&T UNIX documentation.  I don't know why the original author of this comment thought that interfaces "shouldn't" have names that begin with digits, but this has been legal on many systems for decades.  Future maintainers of this code shouldn't be led to believe that this is an invalid configuration that they don't need to handle (which is possibly how we got here in the first place).

**Comment by guyharris** [comment]
> but this has been legal on many systems for decades

"Shouldn't" can either be prescriptive or descriptive; in this case, it's descriptive, because, until Linux or whoever decided to start giving purely-numerical names to interfaces, no UN\*X systems actually did have purely-numerical names for interfaces - not 4.xBSD, not SunOS, not HP-UX, not AIX, not Tru64 UNIX, not any of the BSDs descended from 4.x, etc..

Given that, *now*, there are UN\*X systems that have all-numeric interface names, the comment can be removed, but it's not as if all-numeric interface names have been a thing for decades.

File: tcpdump.c [hunk 1]
```diff
@@ -1371,8 +1405,29 @@ main(int argc, char **argv)
 #endif /* _WIN32 */
 #ifdef HAVE_PCAP_CREATE
 		pd = pcap_create(device, ebuf);
-		if (pd == NULL)
+		if (pd == NULL) {
```
Comment (guyharris) [inline tcpdump.c]: You are assuming here that `pcap_create()` will fail if an interface name is supplied that does not correspond to an interface on the system.  It is not guaranteed to do so on all platforms.  You have to wait for `pcap_activate()` to fail with `PCAP_ERROR_NO_SUCH_DEVICE` (other errors mean there *is* such a device but you can't open it).

This also doesn't handle the case of tcpdump with an older version of libpcap without `pcap_create()`.

**Comment by nbastin** [comment]
> "Shouldn't" can either be prescriptive or descriptive; in this case, it's descriptive, because, until Linux or whoever decided to start giving purely-numerical names to interfaces, no UN*X systems actually did have purely-numerical names for interfaces - not 4.xBSD, not SunOS, not HP-UX, not AIX, not Tru64 UNIX, not any of the BSDs descended from 4.x, etc..

My comment very clearly said it was "legal" - and still is. Leading someone in the future to believe they can rely on such a restriction would create more problems.  Neither POSIX nor AT&T ever said you shouldn't do it, as X.25 PAD interfaces commonly did - and probably others before that - which makes "shouldn't" in the comment prescriptive.  Interfaces have long been named numerically, the fact that PARC didn't do it for ethernet doesn't make it less valid.  Certainly it's even (potentially) more common today than it was in the past (as I believe that under Linux this was in fact not legal for quite some time, while it was always legal in most other UNIX derivatives).

Common practice as known to a developer in their experience is not a spec, it's just one practice.

**Comment by guyharris** [comment]
> which makes "shouldn't" in the comment prescriptive

What would make it prescriptive or descriptive would be the intent of the person who wrote it; the person who wrote it meant it descriptively, which makes it descriptive.

> Interfaces have long been named numerically

On which operating systems that supported libpcap and tcpdump were the names used in libpcap calls, and on the tcpdump command line, numerical?  If we look at [an older version of the FreeBSD `net/ifvar.h`](https://svnweb.freebsd.org/base/head/sys/net/if_var.h?revision=21259&view=markup), a `struct ifnet` has `if_name` and `if_unit` members; the convention was to have `if_name` be a short name for a particular model of network adapter and `if_unit` be a 0-based ordinal for the particular interface in the set of all interfaces of that model, with `if_name` not being purely numerical (or even ending with a digit), as that makes it hard to tell where the name ends and the unit number begins.  Many other UN\*Xes followed BSD in that convention; Linux and OS X used a similar convention, but with `if_name` not being different for different adapter types, and with the Linux loopback interface not having a unit number.

Interfaces may have been named numerically in *other* contexts, but, for tcpdump, the only context that matters is the "argument to the `-i` flag" context.

**Comment by guyharris** [comment]
Fixed instead by 0bc8e0e01ef0b14ed8663f37ce224f35f5fb496b.

**Comment by guyharris** [comment]
> Do we need to modify tcpdump man page to explain this behavior ?

Done in d4876278a77ed6576073cc6fd81d2ec43eb85733.

**Comment by bijalthanawala** [comment]
Thanks guy. Reviewing it.

**Pull request closed without merging**